pkg/discovery/worker/compliance/affinity_commodity_manager.go

```diff
@@ -0,0 +1,124 @@
+	var accessCommsSold []*proto.CommodityDTO
+	var accessCommsBought []*proto.CommodityDTO
+	for _, term := range nodeSelectorTerms {
+		commSold, commBought, err := acm.getCommoditySoldAndBought(term.String())
+		if err != nil {
+			return nil, nil, err
+		}
+		accessCommsSold = append(accessCommsSold, commSold)
+		accessCommsBought = append(accessCommsBought, commBought)
+	}
+	return accessCommsSold, accessCommsBought, nil
+}
+
+func (acm *AffinityCommodityManager) GetAccessCommoditiesForPodAffinityAntiAffinity(podAffinityTerm []api.PodAffinityTerm) ([]*proto.CommodityDTO, []*proto.CommodityDTO, error) {
+	var accessCommsSold []*proto.CommodityDTO
+	var accessCommsBought []*proto.CommodityDTO
+	for _, term := range podAffinityTerm {
+		commSold, commBought, err := acm.getCommoditySoldAndBought(term.String())
+		if err != nil {
+			return nil, nil, err
+		}
+		accessCommsSold = append(accessCommsSold, commSold)
+		accessCommsBought = append(accessCommsBought, commBought)
+	}
+	return accessCommsSold, accessCommsBought, nil
+}
+
+func (acm *AffinityCommodityManager) getCommoditySoldAndBought(termString string) (*proto.CommodityDTO, *proto.CommodityDTO, error) {
+	key, err := generateKey(termString)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to generate hash: %s", err)
+	}
+	commSold, err := acm.getCommoditySold(key)
+	if err != nil {
+		// return immediately even if only one failed.
+		return nil, nil, fmt.Errorf("failed to get accessCommodityDTO sold based on given expressions: %s", err)
+	}
+
+	commBought, err := acm.getCommodityBought(key)
+	if err != nil {
+		// return immediately even if only one failed.
+		return nil, nil, fmt.Errorf("failed to get accessCommodityDTO bought based on given expressions: %s", err)
+	}
+
+	return commSold, commBought, nil
+}
+
+func (acm *AffinityCommodityManager) getCommoditySold(key string) (*proto.CommodityDTO, error) {
+	commodityDTO, exist := acm.commoditySoldStore[key]
+	if exist {
+		return commodityDTO, nil
+	}
+	commodityDTO, err := sdkbuilder.NewCommodityDTOBuilder(proto.CommodityDTO_VMPM_ACCESS).
+		Key(key).
+		Capacity(accessCommodityDefaultCapacity).
+		Create()
+	if err != nil {
+		return nil, err
+	}
+
+	// put into store.
+	acm.commoditySoldStore[key] = commodityDTO
+
+	return commodityDTO, nil
+}
+
+func (acm *AffinityCommodityManager) getCommodityBought(key string) (*proto.CommodityDTO, error) {
+	commodityDTO, exist := acm.commodityBoughtStore[key]
+	if exist {
+		return commodityDTO, nil
+	}
+	commodityDTO, err := sdkbuilder.NewCommodityDTOBuilder(proto.CommodityDTO_VMPM_ACCESS).
+		Key(key).
+		Create()
+	if err != nil {
+		return nil, err
+	}
+
+	// put into store.
+	acm.commodityBoughtStore[key] = commodityDTO
+
+	return commodityDTO, nil
+}
+
+func generateKey(termString string) (string, error) {
+	hashCode, err := hashstructure.Hash(termString, nil)
```

Comment:
The key of the commodity is generated in every termString? Will that create dup commodity that represent the same rule?